primitives/primproc/bppsendthread.cpp

```diff
@@ -53,29 +51,38 @@ BPPSendThread::~BPPSendThread()
 {
     boost::mutex::scoped_lock sl(msgQueueLock);
     boost::mutex::scoped_lock sl2(ackLock);
+    boost::mutex::scoped_lock sl3(respondLock);
     die = true;
     queueNotEmpty.notify_one();
     okToSend.notify_one();
+    okToRespond.notify_one();
     sl.unlock();
     sl2.unlock();
+    sl3.unlock();
     runner.join();
 }
 
-bool BPPSendThread::okToProceed()
-{
-    // keep the queue size below the 100 msg threshold & below the 25MB mark,
-    // but at least 2 msgs so there is always 1 ready to be sent.
-    return ((msgQueue.size() < sizeThreshold && currentByteSize < maxByteSize)
-            || msgQueue.size() < 3) && !die;
-}
-
 void BPPSendThread::sendResult(const Msg_t& msg, bool newConnection)
 {
+    // Wait for the queue to empty out a bit if it's stuffed full
+    if (sizeTooBig())
+    {
+        boost::mutex::scoped_lock sl1(respondLock);
```

Comment:
I personally don't like using boost where std:: elements can be used.

Reply:
Can you replace boost::mutex with std::mutex?

Reply:
NP